File: python_code/raw_python_impls/fo_grouping.py
"""
Grouping algorithm for FOs.
Based on: https://doi.org/10.1109/TKDE.2015.2445755
"""
from dataclasses import dataclass
import math
from typing import Callable

import numpy as np

@dataclass
class Cell:
    """
    Cell of grouping grid
    """
    x_min: int
    y_min: int
    objects: list[object]
    x_centroid: float = None
    y_centroid: float = None

@dataclass
class Group:
    """
    Group.
    """
    cells: list[Cell]

def pre_group(
        objects: object, x_thresh: float, y_thresh: float, get_x: Callable, get_y: Callable
    ) -> list[Group]:
    """
    Pregrouping. Create unoptimized groups.
    :param objects: Objects to group
    :param x_thresh: Threshhold for first attribute
    :param y_thresh: Threshhold for second attribute
    :param get_x: get_x(object) -> first attribute
    :param get_y: get_y(object) -> second attribute
    """
    # Get populated cells
    populated_cells = {}
    for o in objects:
        key = (get_x(o) // (x_thresh+1), get_y(o) // (y_thresh+1))
        if key in populated_cells:
            populated_cells[key].objects.append(o)
        else:
            populated_cells[key] = Cell(get_x(o), get_y(o), [o])
    # Combine adjacent groups
    group_hash = {}
    groups = []
    for base_key, cell in populated_cells.items():
        # Go through adjacent cells
        for i, j in [(-1, 0), (1,0), (0,0), (0, 1), (0, -1)]:
            key = (base_key[0] + i, base_key[1] + j)
            if key in group_hash:
                group = group_hash[key]
                group_hash[base_key] = group
                group.cells.append(cell)
                break
        else:
            new_group = Group([cell])
            group_hash[base_key] = new_group
            groups.append(new_group)
    return groups

def get_cell_centroid(cell: Cell, get_x: Callable, get_y: Callable ) -> tuple[float, float]:
    """
    Get centeroid of a cell.
    :param cell: Cell
    :param get_x: get_x(object) -> first attribute
    :param get_y: get_y(object) -> second attribute
    """
    objects = [o for o in cell.objects]
    x_min = get_x( min(objects, key=get_x) )
    x_max = get_x( max(objects, key=get_x) )
    y_min = get_y( min(objects, key=get_y) )
    y_max = get_y( max(objects, key=get_y) )
    return x_max - x_min, y_max - y_min

def is_mbr_ok(
        group: Group, x_thresh: float, y_thresh: float, get_x: Callable, get_y: Callable
    ) -> bool:
    """
    Check if minimal bouding recantangle of the group does not violates the threshholds.
    :param group: Group
    :param x_thresh: Threshhold for first attribute
    :param y_thresh: Threshhold for second attribute
    :param get_x: get_x(object) -> first attribute
    :param get_y: get_y(object) -> second attribute
    """
    objects = [o for c in group.cells for o in c.objects]
    # x-Coordinate
    te_min = get_x( min(objects, key=get_x) )
    te_max = get_x( max(objects, key=get_x) )
    # y-Coordinate
    tf_min = get_y( min(objects, key=get_y) )
    tf_max = get_y( max(objects, key=get_y) )
    # Check
    te_ok = (te_max - te_min) <= x_thresh
    tf_ok = (tf_max - tf_min) <= y_thresh
    return te_ok and tf_ok

def cluster_distance(group_a, group_b):
    """
    Maximum linkage
    """
    max_distance = 0
    for c_a in group_a.cells:
        for c_b in group_b.cells:
            distance = math.sqrt(
                (c_a.x_centroid-c_b.x_centroid)**2 + (c_a.y_centroid-c_b.y_centroid)**2
            )
            if distance > max_distance:
                max_distance = distance
    return max_distance

def cluster_hierarch(
        cells: list[Cell], x_thresh: float, y_thresh: float, get_x: Callable, get_y: Callable
    ) -> list[Group]:
    """
    Cluster the cells into groups.
    :param cells: Cells
    :param x_thresh: Threshhold for first attribute
    :param y_thresh: Threshhold for second attribute
    :param get_x: get_x(object) -> first attribute
    :param get_y: get_y(object) -> second attribute
    """
    groups = [Group([c]) for c in cells]
    for c in cells:
        x, y = get_cell_centroid(c, get_x, get_y)
        c.x_centroid = x
        c.y_centroid = y
    while len(groups) > 1:
        # Find minimum distance between to groups
        min_pair = None
        min_distance = np.inf
        n = len(groups)
        for i in range(n):
            for j in range(i+1, n):
                distance = cluster_distance(groups[i], groups[j])
                if distance < min_distance:
                    min_distance = distance
                    min_pair = (i, j)
        # Merge groups
        (i, j) = min_pair
        group_a = groups[i]
        group_b = groups[j]
        new_group = Group(group_a.cells + group_b.cells)
        if is_mbr_ok(new_group, x_thresh, y_thresh, get_x, get_y):
            groups.remove(group_a)
            groups.remove(group_b)
            groups.append(new_group)
        else:
            break
    return groups

def optimize_groups(
        groups: list[Group],  x_thresh: float, y_thresh: float, get_x: Callable, get_y: Callable
    ) -> list[Group]:
    """
    Optimize grouping. Only splitting needs to be done.
    All adjacent groups are already merged in pre_group().
    :param groups: Groups
    :param x_thresh: Threshhold for first attribute
    :param y_thresh: Threshhold for second attribute
    :param get_x: get_x(object) -> first attribute
    :param get_y: get_y(object) -> second attribute
    """
    opt_groups = []
    for group in groups:
        if not is_mbr_ok(group, x_thresh, y_thresh, get_x, get_y):
            opt_groups.extend(cluster_hierarch(group.cells, x_thresh, y_thresh, get_x, get_y))
        else:
            opt_groups.append(group)
    return opt_groups

File: src/fo_grouping.rs
use std::collections::HashMap;

use indexmap::IndexMap;

#[derive(Clone)]
pub struct Cell<T> {
    pub objects: Vec<T>,
    x_centroid: i32,
    y_centroid: i32
}

#[derive(Clone)]
pub struct Group<T> {
    pub cells: Vec<Cell<T>>
}

impl <T> Cell<T> {
    fn update_centroid(&mut self, get_x: fn(&T) -> i32, get_y: fn(&T) -> i32) {
        let xmin = self.objects.iter().map(get_x).min().unwrap();
        let xmax = self.objects.iter().map(get_x).max().unwrap();
        let ymin = self.objects.iter().map(get_y).min().unwrap();
        let ymax = self.objects.iter().map(get_y).max().unwrap();
        self.x_centroid = xmax - xmin;
        self.y_centroid = ymax - ymin;
    }    
}

pub fn pre_group<T>(objects: Vec<T>, x_thresh: i32, y_thresh: i32, get_x: fn(&T) -> i32, get_y: fn(&T) -> i32) -> Vec<Group<T>> {
    // Get populated cells
    let mut populated_cells: IndexMap<(i32, i32), Cell<T>> = indexmap::IndexMap::new(); // Use indexmap for consistency with Python version.
    for o in objects {
        let x_o = get_x(&o);
        let y_o = get_y(&o);
        let key = (x_o / (x_thresh + 1), y_o / (y_thresh + 1));
        if populated_cells.contains_key(&key) {
            populated_cells.get_mut(&key).unwrap().objects.push(o);
        } else {
            let new_cell = Cell {objects: vec![o], x_centroid: -1, y_centroid: -1};
            populated_cells.insert(key, new_cell);
        }
    }
    // Combine adjacent groups
    let mut combined_groups: Vec<Group<T>> = Vec::new();
    let mut group_hash: HashMap<(i32, i32), usize> = HashMap::new();
    let key_shifts = [(-1, 0), (1,0), (0,0), (0, 1), (0, -1)];
    for (base_key, cell) in populated_cells {
        // Go through adjacent cells
        let used_key = key_shifts.iter()
            .map(|(i, j)| (base_key.0 + i, base_key.1 + j))
            .find(|x| group_hash.contains_key(&x));
        match used_key {
            Some(key) => {
                let pntr = *group_hash.get(&key).unwrap();
                group_hash.insert(base_key, pntr);
                combined_groups[pntr].cells.push(cell);
            }
            None => {
                let new_group = Group{cells: vec![cell]};
                let pntr = combined_groups.len();
                combined_groups.push(new_group);
                group_hash.insert(base_key, pntr);
            }
        }
    }
    return combined_groups;
}

fn mbr_ok<'a, I, T>(vals: I, x_thresh: i32, y_thresh: i32, get_x: fn(&T) -> i32, get_y: fn(&T) -> i32) -> bool 
where
    I: Iterator<Item = &'a Cell<T>>, I: Clone, T: 'a
{
    let xmin = vals.clone().flat_map(|x| &x.objects).map(get_x).min().unwrap();
    let xmax = vals.clone().flat_map(|x| &x.objects).map(get_x).max().unwrap();
    let ymin = vals.clone().flat_map(|x| &x.objects).map(get_y).min().unwrap();
    let ymax = vals.clone().flat_map(|x| &x.objects).map(get_y).max().unwrap();
    return (xmax - xmin <= x_thresh) && (ymax - ymin <= y_thresh)
}

fn maximum_linkage_distance<T>(grp_a: &Group<T>, grp_b: &Group<T>) -> f64 {
    let mut max_distance = 0.0;
    for c_a in &grp_a.cells {
        for c_b in &grp_b.cells {
            let distance = f64::sqrt(
                ((c_a.x_centroid - c_b.x_centroid).pow(2) + (c_a.y_centroid - c_b.y_centroid).pow(2)) as f64
            );
            max_distance = f64::max(distance, max_distance);
        }
    }
    return max_distance
}

fn hierachical_clustering<T>(cells: Vec<Cell<T>>, x_thresh: i32, y_thresh: i32, get_x: fn(&T) -> i32, get_y: fn(&T) -> i32) -> Vec<Group<T>> {
    let mut cells = cells;
    for c in &mut cells {
        c.update_centroid(get_x, get_y);
    }
    let mut groups: Vec<Group<T>> = cells.into_iter().map(|x| Group{cells: vec![x]}).collect();
    while groups.len() > 1 {
        // Find closest pair
        let mut min_pair: Option<(usize, usize)> = None;
        let mut min_distance = f64::INFINITY;
        for i in 0..groups.len() {
            for j in i+1..groups.len() {
                let distance = maximum_linkage_distance(&groups[i], &groups[j]);
                if distance < min_distance {
                    min_distance = distance;
                    min_pair = Some((i, j));
                }
            }
        }
        // Merge pair
        let (i, j) = min_pair.unwrap();
        let cell_iter = groups[i].cells.iter().chain(groups[j].cells.iter());
        if mbr_ok(cell_iter, x_thresh, y_thresh, get_x, get_y) {
            let group_a = groups.remove(j); // j > i
            let group_b = groups.remove(i);
            let mut cells = group_a.cells;
            cells.extend(group_b.cells);
            let new_group = Group{ cells };
            groups.push(new_group);
        } else {
            break;
        }
    }
    return groups
}

pub fn optimize_groups<T>(groups: Vec<Group<T>>, x_thresh: i32, y_thresh: i32, get_x: fn(&T) -> i32, get_y: fn(&T) -> i32) -> Vec<Group<T>> {
    let mut opt_groups = Vec::new();
    for group in groups {
        if !mbr_ok(group.cells.iter(), x_thresh, y_thresh, get_x, get_y) {
            let clustered = hierachical_clustering(group.cells, x_thresh, y_thresh, get_x, get_y);
            opt_groups.extend( clustered );
        } else {
            opt_groups.push(group)
        }
    }
    return opt_groups
}  